acme/acme.go

```diff
@@ -328,29 +355,44 @@ func (a *ACME) CreateLocalConfig(tlsConfig *tls.Config, checkOnDemandDomain func
 func (a *ACME) getCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
 	domain := types.CanonicalDomain(clientHello.ServerName)
 	account := a.store.Get().(*Account)
-	//use regex to test for wildcard certs that might have been added into TLSConfig
-	for k := range a.TLSConfig.NameToCertificate {
-		selector := "^" + strings.Replace(k, "*.", "[^\\.]*\\.?", -1) + "$"
-		match, _ := regexp.MatchString(selector, domain)
-		if match {
-			return a.TLSConfig.NameToCertificate[k], nil
-		}
-	}
+
+	// Check for existing challenge cert
+	log.Debugf("Checking ACME challenge for %s", domain)
 	if challengeCert, ok := a.challengeProvider.getCertificate(domain); ok {
-		log.Debugf("ACME got challenge %s", domain)
+		log.Debugf("Returning ACME challenge cert for %s", domain)
 		return challengeCert, nil
 	}
+
+	// Check for existing ACME domain cert
+	log.Debugf("Checking ACME domain cert for %v", domain)
 	if domainCert, ok := account.DomainsCertificate.getCertificateForDomain(domain); ok {
-		log.Debugf("ACME got domain cert %s", domain)
+		log.Debugf("Returning ACME domain cert for %s", domain)
 		return domainCert.tlsCert, nil
 	}
+
 	if a.OnDemand {
 		if a.checkOnDemandDomain != nil && !a.checkOnDemandDomain(domain) {
 			return nil, nil
 		}
-		return a.loadCertificateOnDemand(clientHello)
+		cert, err := a.loadCertificateOnDemand(clientHello)
+		if cert != nil && err == nil {
+			return cert, nil
+		}
+		log.Errorf("On demand certicate retrival failed for %v due to %v", domain, err)
 	}
-	log.Debugf("ACME got nothing %s", domain)
+
+	// Check if we have a wildcard cert into TLSConfig that could be used
+	log.Debugf("Checking wildcard certificate matching for %v", domain)
+	for k := range a.TLSConfig.NameToCertificate {
```

Comment:
Many thanks for you reply.

As you just said, the Part1 of your response changes deeply the current behavior if the `IgnoreFilter` field is used to filter domains checked by LE.

But I believe the suggestion we did (use `IgnoreFilter` to filter (sub)domains checked by a wildcard certificate) allows to respect the 2 use cases you described :
1. I have a signed wildcard and I want to use it instead of LE certificate : I don't use `IgnoreFilter` field, the behavior will remain the same as now.
2. I have a self-signed wildcard and I prefer to use LE for any (all?) (sub)domains : I add the list of (sub)domains which haven't to be checked by the wildcard certificate into the `IgnoreFilter` field and the LE certificate will be used.

@CyrilPeponnet  Are you agree with this?

cc/  @dtomcej @emilevauge @Juliens @ldez